faiss/impl/ResidualQuantizer.cpp

```diff
@@ -381,13 +409,128 @@ void ResidualQuantizer::train(size_t n, const float* x) {
         qnorm.add(clus.k, clus.centroids.data());
     }
 
-    is_trained = true;
-
     if (!(train_type & Skip_codebook_tables)) {
         compute_codebook_tables();
     }
 }
 
+float ResidualQuantizer::retrain_AQ_codebook(size_t n, const float* x) {
```

Comment:
That looks great! If I understand correctly, it is similar to `LocalSearchQuantizer::update_codebooks`. Could we move this function to AdditiveQuantizer?

Reply:
Ah right, there is some redundancy with LSQ. LMC which function is better to keep. Probably needs to be merged because yours has a double version but mine supports multiple nb bits per quantizer.

Reply:
Sorry, I am going to integrate the PR as it is now because I need it quickly. 
For reference the differences are: 
- both float + double support for the LocalSearchQuantizer version
- variable nb bits support for the ResidualQuantizer version
- degeneracy (presumably due to missing example codes in the training data): 
* LSQ --> add lamdba * diagonal to linear system matrix 
* RQ --> explicit estimate of rank of the matrix + clip SVD decomposition to this rank 
The merits of both solutions should be compared when merging.